samples/subgroup-scan.cpp

```diff
@@ -0,0 +1,208 @@
+    cgh.parallel_for<kernel_name<T, Op, class scan_segments>>(
+        sycl::nd_range<1>(input_size, wgroup_size), [=](sycl::nd_item<1> item) {
+          size_t gid = item.get_global_linear_id();
+
+          auto sub_group = item.get_sub_group();
+          auto scan_res = inclusive_scan_over_group(sub_group, data[gid], Op{});
+          if (sub_group.get_local_id() == sub_group.get_local_range() - 1) {
+            temp[sub_group.get_group_id()[0]] = scan_res;
+          }
+          item.barrier(sycl::access::fence_space::local_space);
+          for (auto i = 1u; i < sub_group.get_group_range()[0]; i++) {
+            scan_res += sub_group.get_group_id()[0] >= i ? temp[i - 1] : 0;
+          }
+          data[gid] = scan_res;
+        });
+  });
+
+  // At this point we have computed the inclusive scans of this many segments.
+  size_t n_segments = input_size / wgroup_size;
+  if (n_segments == 1) {
+    // If all of the data is in one segment, we're done.
+    return;
+  }
+
+  // Store the last element of each segment in a temporary buffer
+  sycl::buffer<T, 1> ends{sycl::range<1>(n_segments)};
+  q.submit([&](sycl::handler& cgh) {
+    auto scans = in.template get_access<mode::read>(cgh);
+    auto elems = ends.template get_access<mode::discard_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, class copy_ends>>(
+        sycl::range<1>(n_segments), [=](sycl::item<1> item) {
+          auto id = item.get_linear_id();
+          // Offset into the last element of each segment.
+          elems[item] = scans[(id + 1) * wgroup_size - 1];
+        });
+  });
+
+  // Recursively scan the array of last elements.
+  par_scan<T, Op>(ends, q);
+
+  // Add the results of the scan to each segment.
+  q.submit([&](sycl::handler& cgh) {
+    auto ends_scan = ends.template get_access<mode::read>(cgh);
+    auto data = in.template get_access<mode::read_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, class add_ends>>(
+        // Work with one less work-group, since the first segment is correct.
+        sycl::nd_range<1>(input_size - wgroup_size, wgroup_size),
+        [=](sycl::nd_item<1> item) {
+          auto group = item.get_group_linear_id();
+          auto off_gid = item.get_global_linear_id() + wgroup_size;
+          data[off_gid] = Op{}(data[off_gid], ends_scan[group]);
+        });
+  });
+}
+
+/* Tests the scan with an addition operation, which is its most common use.
+ * Returns 0 if successful, a nonzero value otherwise. */
+int test_sum(sycl::queue& q) {
+  constexpr size_t size = 8192;
+
+  std::vector<int32_t> in(size);
+  std::iota(in.begin(), in.end(), 1);
+
+  std::vector<int32_t> sum(in.size());
+  {
+    sycl::buffer<int32_t, 1> buf(sycl::range<1>(in.size()));
+    buf.set_final_data(sum.data());
+    q.submit([&](sycl::handler& cgh) {
+      auto acc = buf.get_access<sycl::access::mode::write>(cgh);
+      cgh.copy(in.data(), acc);
+    });
+
+    par_scan<int32_t, sycl::experimental::plus<int32_t>>(buf, q);
```

Comment:
Experimental?

Reply:
Ah, I don't know!  I assume the updated version works though.

Reply:
It does indeed. @hjabird should I prefer `experimental::` or just `sycl::`?

Reply:
`sycl`.

Reply:
Thanks, I'll leave it is as now then!